java/code/src/com/redhat/rhn/common/hibernate/ReportDbHibernateFactory.java

```diff
@@ -0,0 +1,873 @@
+/*
+ * Copyright (c) 2021 SUSE LLC
+ *
+ * This software is licensed to you under the GNU General Public License,
+ * version 2 (GPLv2). There is NO WARRANTY for this software, express or
+ * implied, including the implied warranties of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. You should have received a copy of GPLv2
+ * along with this software; if not, see
+ * http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt.
+ *
+ * Red Hat trademarks are not licensed under GPLv2. No permission is
+ * granted to use or replicate Red Hat trademarks that are incorporated
+ * in this software or its documentation.
+ */
+package com.redhat.rhn.common.hibernate;
+
+import com.redhat.rhn.common.db.DatabaseException;
+import com.redhat.rhn.common.db.datasource.CallableMode;
+import com.redhat.rhn.common.db.datasource.DataResult;
+import com.redhat.rhn.common.db.datasource.ModeFactory;
+import com.redhat.rhn.common.db.datasource.SelectMode;
+
+import org.apache.commons.collections.ListUtils;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.log4j.Logger;
+import org.hibernate.Hibernate;
+import org.hibernate.HibernateException;
+import org.hibernate.Session;
+import org.hibernate.query.Query;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.UnsupportedEncodingException;
+import java.sql.Blob;
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.BinaryOperator;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+import javax.persistence.FlushModeType;
+
+/**
+ * HibernateFactory - Helper superclass that contains methods for fetching and
+ * storing Objects from the DB using Hibernate.
+ * <p>
+ * Abstract methods define what the subclass must implement to determine what is
+ * specific to that Factory's instance.
+ */
+public class ReportDbHibernateFactory {
```

Comment:
If i understand this correctly this is just a copy of HibernateFactory with a few adjustments. Is there a plan to have most of the code shared later?

Reply:
It would be indeed better, but the problem is that HibernateFactory has both static and instance methods, it is extended by multiple factories and all of them share statically the same session. Currently HibernateFactory has a static ConnectionManager, while ReportDbHibernateFactory has an instance field for that. This might be problematic if we want to extend it or have multiple instances. Not sure though if it's something we will have in our use case.

Also, I think we can get rid of most of the duplication if we plan to use ReportDbHibernateFactory only need to execute mode queries.